include/circt/Dialect/RTL/RTLTypes.td

```diff
@@ -86,3 +86,66 @@ class InOutElementConstraint<string value, string inoutValue>
   : TypesMatchWith<"type should be element of inout type",
                    inoutValue, value,
                    "$_self.cast<InOutType>().getElementType()">;
+
+// Adapted from TestTypeDefs.td
+// A parent type for any type which is just a list of fields (e.g. structs,
+// unions).
+def StructType : RTLType<"Struct"> {
+  let summary = "RTL struct type";
+  let description = [{
+    Struct type has value semantics, updates to the struct are logically by 
+    copy.
+  }];
+  let mnemonic = "struct";
+
+  let parameters = (
+    ins
+    // An ArrayRef of something which requires allocation in the storage
+    // constructor.
+    ArrayRefOfSelfAllocationParameter<
+      "::circt::rtl::StructType::FieldInfo",
+      "struct fields">: $elements
+  );
+
+  // Prints the type in this format:
+  //   struct<[{field1Name, field1Type}, {field2Name, field2Type}]>
+  let printer = [{
+    $_printer << "struct" << "<";
+    for (size_t i=0, e = getImpl()->elements.size(); i < e; i++) {
+      const auto& field = getImpl()->elements[i];
+      $_printer << "{" << field.name << "," << field.type << "}";
+      if (i < getImpl()->elements.size() - 1)
+          $_printer << ",";
+    }
```

Comment:
You might use the `llvm::interleaveComma` algorithm to simplify this.  I think something close to:

```
llvm::interleaveComma(getImpl()->elements, $_printer, [&](const auto &field) {
$_printer << "{" << field.name << "," << field.type << "}";
});
```
would work